mcs/class/System.Xaml/System.Xaml/XamlBackgroundReader.cs

```diff
@@ -105,11 +105,19 @@ public void StartThread (string threadName)
 			if (thread != null)
 				throw new InvalidOperationException ("Thread has already started");
 			thread = new Thread (new ParameterizedThreadStart (delegate {
-				while (do_work && r.Read ()) {
-					q.Writer.WriteNode (r);
+				try {
+					while (do_work && r.Read ()) {
+						q.Writer.WriteNode (r);
+						wait.Set ();
+					}
+					read_all_done = true;
+				}
+				catch (Exception ex)
+				{
 					wait.Set ();
+					read_all_done = true;
```

Comment:
I'd not set read_all_done in this case
Please use `} catch {` instead

Reply:
Updated to not set read_all_done but instead bubble up the exception to Read(), which is required otherwise we'd still be in a deadlock since the calling code calls Read() in a loop if IsEof == false.